cdap-api/src/main/java/co/cask/cdap/api/data/batch/Input.java

```diff
@@ -0,0 +1,155 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.data.batch;
+
+import co.cask.cdap.api.data.stream.StreamBatchReadable;
+
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Defines input to a MapReduce job.
+ */
+public abstract class Input {
+
+  private final String name;
+
+  private Input(String name) {
+    this.name = name;
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   */
+  public static Input of(String datasetName) {
+    return of(datasetName, Collections.<String, String>emptyMap());
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param arguments the arguments to use when instantiating the dataset
+   */
+  public static Input of(String datasetName, Map<String, String> arguments) {
+    return of(datasetName, arguments, null);
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param splits the data selection splits
+   */
+  public static Input of(String datasetName, List<Split> splits) {
+    return of(datasetName, Collections.<String, String>emptyMap(), splits);
+  }
+
+  /**
+   * Returns an Input defined by a dataset.
+   *
+   * @param datasetName the name of the input dataset
+   * @param arguments the arguments to use when instantiating the dataset
+   * @param splits the data selection splits
+   */
+  public static Input of(String datasetName, Map<String, String> arguments, List<Split> splits) {
+    return new DatasetInput(datasetName, arguments, splits);
+  }
+
+  /**
+   * Returns an Input defined by an InputFormatProvider.
+   *
+   * @param inputName the name of the input
+   */
+  public static Input of(String inputName, InputFormatProvider inputFormatProvider) {
+    return new InputFormatProviderInput(inputName, inputFormatProvider);
+  }
+
+  /**
+   * Returns an Input defined by a stream.
+   *
+   * @param streamBatchReadable specifies the stream to be used as input
+   */
+  public static Input of(StreamBatchReadable streamBatchReadable) {
+    return new StreamInput(streamBatchReadable);
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * An implementation of {@link Input}, which defines a {@link co.cask.cdap.api.dataset.Dataset} as an input.
+   */
+  public static class DatasetInput extends Input {
+
+    private final Map<String, String> arguments;
+    private final List<Split> splits;
+
+    public DatasetInput(String name, Map<String, String> arguments, List<Split> splits) {
+      super(name);
+      this.arguments = arguments;
+      this.splits = splits;
```

Comment:
How can we do that - if we make it Iterable (https://github.com/caskdata/cdap/pull/5276#discussion_r55805171)?
Better to make it Collection then, or leave it List?